rfcs/003-nix-nickel.md

```diff
@@ -0,0 +1,460 @@
+both.
+
+This solution requires to reimplement Nix builtins in Nickel (the nickel-nix
+compatibility layer) . This is an interesting milestone in itself, because even
+without a Nix-to-Nickel compiler, the compatibility layer would already make
+writing derivations in pure Nickel possible.
+
+- TODO: what about dynamic imports?
+- TODO: what about using a Nickel pkgs from Nix?
+
+#### Using Nixpkgs in the PARM
+
+Using a package from Nixpkgs in the PARM with a Nix-to-Nickel compiler would be
+straightforward, as we could evaluate anything to a derivation whenever needed.
+One would need to use Nixpkgs functions and idioms to do e.g. overrides though,
+but this seems pretty hard to avoid, as a systematic translation from the
+Nixpkgs model to the PARM doesn't seem trivial, if even doable.
+
+<!-- TODO: add examples of such interactions -->
+
+### Supporting string contexts
+
+#### Nix-style
+
+Ideally, we would like to have a behavior similar to Nix string contexts to
+automatically track dependencies in a transparent way (you don't have to think
+about it in Nix, using strings in a natural way).
+
+There is a tension between endowing Nickel with specific features for the Nix
+use-case and keeping it a general and versatile language for configuration.
+
+One solution is to support string contexts in Nickel exactly as in Nix. But that
+would do feel _ad hoc_. Other use-cases would benefit from a similar but more
+flexible mechanism for automatic dependency tracking. For example, in Terraform,
+interpolated expressions can refer to values that are only known after certain
+resources have been deployed. The Terraform evaluator thus uses a similar
+mechanism to elaborate a deployment plan from string metadata. Using Nickel for
+Terraform would need to replicate this mechanism in some way.
+
+We propose to adopt a more generic mechanism for strings with context, which
+behavior can be parametrized. Strings become (conceptually) pairs `(s, ctxt)`
+where `ctxt` is an arbitrary data structure with additional structure:
+
+```Rust
+combine: Ctxt -> Ctxt -> Ctxt,
+pure: Str -> Ctxt,
+pure_exp: Expr -> Ctxt,
+```
+
+The above functions could either be provided by user code or directly as an
+interpreter plug-in.
+
+- Existing standard strings would be equivalent to having a trivial context
+  `null` and the obvious corresponding trivial structure.
+- Custom contexts would be introduced by specific string delimiters, such as
+    `nix%" %{pkgs.hello}/bin/hello"%`
+- Different kind of strings are incompatible, to avoid accidentally forgetting
+  or losing contexts. Also, we don't know a priori how to convert one context to
+  another.
+- There would be a contract to distinguish the different kind of
+  strings in order enforce the usage of e.g. Nix style contexts for writing
+  Nickel for Nix. We want to avoid users loosing or missing context unknowingly.
+
+#### G-exps
+
+Alternative: something like `g-exp`. No magic, no extension, but less ergonomic.
```

Comment:
What are `g-exprs`? Sounds like something Scheme-like (ala s-expression / fexprs). Perhaps a reference and an example of how it is less ergonomic would be helpful.

Reply:
Those are the Guix equivalent of string contexts, somehow: [manual page](https://guix.gnu.org/manual/en/html_node/G_002dExpressions.html). At this point this paragraph is just draft notes, but it'll deserve a proper link indeed once finalized. At a quick glance, it's less ergonomic because:

- Use a special syntax to start a g-exp `#~` (gexp)
- Use a special syntax to interpolate other g-exp `#$` (ungexp)

As opposed to Nix where string contexts just work without having the user to even think about it or convert between strings and derivations. On the other hand, the explicitness of g-exp may actually be a good point: arguably, it makes things less magic. But it would still be one additional thing that Nix users don't need to do today. Although, I think what was written at the time is not true: something akin to g-expr in Nickel using standard syntax for strings, like interpolation `%{}`, would still need a bit of magic and an extension mechanism, probably.

Reply:
As I mentioned in IRL, my thinking is that the ideal solution is to have overloaded and customisable interpolated strings. Some languages do this such as Scala and Haskell (via quasiquotes and TH). In Scala it requires a prefix. Haskell doesn't have builtin interpolation, so quasiquotation kind of _is_ the way :man_shrugging:. I experimented a little bit with the the support in Idris the other night. It seems doable. Though I seem to have hit a limitation, misunderstanding, or bug in Idris.

Reply:
In Idris, for instance, you can overload the standard string interpolation syntax. It should be possible to make something like this work:

```idris
postInstall : PkgExp
postInstall = """
    substituteInPlace \{out}/bin/my-script \\
      --replace /usr/bin/python \{python}/bin/python \\
      --replace /usr/bin/perl "\{perl}/bin/perl -I \{libfoo}"
  """
```

I'm not fond of that backslash as it evokes the idea of a "lambda". This is simply how string interpolation is defined in Idris (strange as `$` was discarded because of it's special meaning).

In [this short experiment](https://gist.github.com/steshaw/b3916d42f49f5b4e4ecc2922513cd7ad#file-pkg-idr-L50-L57), `pkg3a` should be equivalent to the desugared `pkg3` but I ran into a snag.